apisix/healthcheck_manager.lua

```diff
@@ -0,0 +1,288 @@
+        return nil
+    end
+    core.log.warn("creating healthchecker for upstream: ", up_conf.resource_key)
+    if not healthcheck then
+        healthcheck = require("resty.healthcheck")
+    end
+
+    core.log.warn("creating new healthchecker for ", up_conf.key)
+
+    local checker, err = healthcheck.new({
+        name = get_healthchecker_name(up_conf),
+        shm_name = healthcheck_shdict_name,
+        checks = up_conf.checks,
+        events_module = events:get_healthcheck_events_modele(),
+    })
+
+    if not checker then
+        core.log.error("failed to create healthcheck: ", err)
+        return nil
+    end
+
+    -- Add target nodes
+    local host = up_conf.checks and up_conf.checks.active and up_conf.checks.active.host
+    local port = up_conf.checks and up_conf.checks.active and up_conf.checks.active.port
+    local up_hdr = up_conf.pass_host == "rewrite" and up_conf.upstream_host
+    local use_node_hdr = up_conf.pass_host == "node" or nil
+
+    for _, node in ipairs(up_conf.nodes) do
+        local host_hdr = up_hdr or (use_node_hdr and node.domain)
+        local ok, err = checker:add_target(node.host, port or node.port, host,
+                                        true, host_hdr)
+        if not ok then
+            core.log.error("failed to add healthcheck target: ", node.host, ":",
+                          port or node.port, " err: ", err)
+        end
+    end
+
+    return checker
+end
+
+
+function _M.fetch_checker(resource_path, resource_ver)
+    local working_item = _M.working_pool[resource_path]
+    if working_item and working_item.version == resource_ver then
+        return working_item.checker
+    end
+
+    if _M.waiting_pool[resource_path] == resource_ver then
+        return nil
+    end
+
+    -- Add to waiting pool with version
+    core.log.info("adding ", resource_path, " to waiting pool with version: ", resource_ver)
+    _M.waiting_pool[resource_path] = resource_ver
+    return nil
+end
+
+
+function _M.fetch_node_status(checker, ip, port, hostname)
+    -- check if the checker is valid
+    if not checker or checker.dead then
+        return true
+    end
+
+    return checker:get_target_status(ip, port, hostname)
+end
+
+
+local function add_working_pool(resource_path, resource_ver, checker)
+    _M.working_pool[resource_path] = {
+        version = resource_ver,
+        checker = checker
+    }
+end
+
+local function find_in_working_pool(resource_path, resource_ver)
+    local checker = _M.working_pool[resource_path]
+    if not checker then
+        return nil  -- not found
+    end
+
+    if checker.version ~= resource_ver then
+        core.log.warn("version mismatch for resource: ", resource_path,
+                    " current version: ", checker.version, " requested version: ", resource_ver)
+        return nil  -- version not match
+    end
+    return checker
+end
+
+local timer_create_checker_running = false
```

Comment:
this variable only be used in `init_worker`, maybe we should move it to before `init_worker` function to easy find it's usage.

Reply:
done